test/source/tests/setup.ts

```diff
@@ -715,33 +725,108 @@ AN8G3r5Htj8olot+jm9mIa5XLXWzMNUZgg==
       await PageRecipe.waitForToastToAppearAndDisappear(gmailPage,
         'Could not update keys from EKM due to error: BrowserMsg(processAndStoreKeysFromEkmLocally) sendRawResponse::Error: Some keys could not be parsed');
       await gmailPage.notPresent('@dialog-passphrase');
-      const set13 = await retrieveAndCheckKeys(3);
-      expect(set13.map(entry => entry.id)).to.eql([
-        '392FB1E9FF4184659AB6A246835C0141B9ECF536',
-        'FAFB7D675AC74E87F84D169F00B0115807969D75',
-        '277D1ADA213881F4ABE0415395E783DC0289E2E2'
-      ]);
-      const mainKey13 = KeyUtil.filterKeysByIdentity(set13, [{ family: 'openpgp', id: '392FB1E9FF4184659AB6A246835C0141B9ECF536' }]);
+      const set13 = await retrieveAndCheckKeys(settingsPage, acct, 1, passphrase);
+      expect(set13.map(entry => entry.id)).to.eql(['277D1ADA213881F4ABE0415395E783DC0289E2E2']);
+      const mainKey13 = KeyUtil.filterKeysByIdentity(set13, [{ family: 'openpgp', id: '277D1ADA213881F4ABE0415395E783DC0289E2E2' }]);
       expect(mainKey13.length).to.equal(1);
-      expect(mainKey13[0].lastModified).to.equal(mainKey10[0].lastModified); // no update
+      expect(mainKey13[0].lastModified).to.equal(mainKey12[0].lastModified); // no update
       await gmailPage.close();
       // 13. EKM down, no toast, no passphrase dialog, no updates
-      MOCK_KM_UPDATING_KEY.badRequestError = 'RequestTimeout';
+      MOCK_KM_UPDATING_KEY[acct] = { badRequestError: 'RequestTimeout' };
       gmailPage = await browser.newPage(t, TestUrls.mockGmailUrl(), undefined, extraAuthHeaders);
       await PageRecipe.noToastAppears(gmailPage);
       await gmailPage.notPresent('@dialog-passphrase');
-      const set14 = await retrieveAndCheckKeys(3);
-      expect(set14.map(entry => entry.id)).to.eql([
-        '392FB1E9FF4184659AB6A246835C0141B9ECF536',
-        'FAFB7D675AC74E87F84D169F00B0115807969D75',
-        '277D1ADA213881F4ABE0415395E783DC0289E2E2'
-      ]);
-      const mainKey14 = KeyUtil.filterKeysByIdentity(set14.map(ki => ki), [{ family: 'openpgp', id: '392FB1E9FF4184659AB6A246835C0141B9ECF536' }]);
+      const set14 = await retrieveAndCheckKeys(settingsPage, acct, 1, passphrase);
+      expect(set14.map(entry => entry.id)).to.eql(['277D1ADA213881F4ABE0415395E783DC0289E2E2']);
+      const mainKey14 = KeyUtil.filterKeysByIdentity(set14.map(ki => ki), [{ family: 'openpgp', id: '277D1ADA213881F4ABE0415395E783DC0289E2E2' }]);
       expect(mainKey14.length).to.equal(1);
       expect(mainKey14[0].lastModified).to.equal(mainKey13[0].lastModified); // no update
       await gmailPage.close();
     }));
 
+    ava.default('put.updating.key@key-manager-choose-passphrase-forbid-storing.flowcrypt.test - updates of key found on key manager via setup page (with passphrase)',
```

Comment:
Do any of the new tests check that keys wouldn't be duplicated in storage when there was an older version of the key previously present? (that we are not storing both original version and updated version of the key alongside each other)

Reply:
All the tests check the total expected number of keys in the storage after each step. If there is an unexpected "extra" key, any of the tests will fail.

Reply:
The check is inside `retrieveAndCheckKeys` that has `expectedKeyCount` as a mandatory parameter.
https://github.com/FlowCrypt/flowcrypt-browser/blob/12a5fb78bda305ae028bd25d2b8aaebecf1cc9af/test/source/tests/setup.ts#L583